packages/contractkit/src/identity/offchain-data-wrapper.ts

```diff
@@ -74,46 +89,93 @@ export default class OffchainDataWrapper {
+      return new FetchError()
     }
-    await this.storageWriter.write(data, dataPath)
-    // TODO: Prefix signing abstraction
-    const sig = await NativeSigner(this.kit.web3.eth.sign, this.self).sign(data)
-    await this.storageWriter.write(sig, dataPath + '.signature')
   }
 }
 
 class StorageRoot {
-  constructor(readonly account: string, readonly address: string) {}
-
-  // TODO: Add decryption metadata (i.e. indicates ciphertext to be decrypted/which key to use)
-  async readAsResult(dataPath: string): Promise<Result<string, OffchainErrors>> {
-    const data = await fetch(this.address + dataPath)
-    if (!data.ok) {
+  constructor(
+    readonly wrapper: OffchainDataWrapper,
+    readonly account: string,
+    readonly address: string
+  ) {}
+
+  async readAndVerifySignature(
+    dataPath: string,
+    buildTypedData: (x: Buffer) => Promise<EIP712TypedData>
+  ): Promise<Result<Buffer, OffchainErrors>> {
+    let dataResponse, signatureResponse
+
+    try {
+      ;[dataResponse, signatureResponse] = await Promise.all([
+        fetch(this.address + dataPath),
+        fetch(this.address + dataPath + '.signature'),
+      ])
+    } catch (error) {
       return Err(new FetchError())
     }
 
-    const body = await data.text()
-
-    const signature = await fetch(this.address + dataPath + '.signature')
-
-    if (!signature.ok) {
+    if (!dataResponse.ok) {
+      return Err(new FetchError())
+    }
+    if (!signatureResponse.ok) {
       return Err(new InvalidSignature())
     }
 
-    // TODO: Compare against registered on-chain signers or off-chain signers
-    const signer = guessSigner(body, await signature.text())
+    const [dataBody, signatureBody] = await Promise.all([
+      dataResponse.arrayBuffer(),
+      signatureResponse.arrayBuffer(),
+    ])
+    const body = Buffer.from(dataBody)
+    const signature = ensureLeading0x(Buffer.from(signatureBody).toString('hex'))
 
-    if (eqAddress(signer, this.account)) {
+    const typedData = await buildTypedData(body)
+    const guessedSigner = guessEIP712TypedDataSigner(typedData, signature)
+
+    if (eqAddress(guessedSigner, this.account)) {
       return Ok(body)
     }
 
-    // The signer might be authorized off-chain
-    // TODO: Only if the signer is authorized with an on-chain key
-    return this.readAsResult(`/account/authorizedSigners/${toChecksumAddress(signer)}`)
-  }
+    const accounts = await this.wrapper.kit.contracts.getAccounts()
+    if (await accounts.isAccount(this.account)) {
+      const signers = await Promise.all([
+        accounts.getVoteSigner(this.account),
+        accounts.getValidatorSigner(this.account),
+        accounts.getAttestationSigner(this.account),
+      ])
+      if (signers.some((signer) => signer === guessedSigner)) {
+        return Ok(body)
+      }
+
+      const authorizedSignerAccessor = new AuthorizedSignerAccessor(this.wrapper)
+      const authorizedSigner = await authorizedSignerAccessor.readAsResult(
```

Comment:
It seems like this may result in an infinite loop because `readAndVerifySignature` is indirectly called by the authorized signature accessor, which will return to here and repeat.

Reply:
Hmm so I added another param, `checkOffchainSignatures: boolean`, but I dunno if that's the best way to do this...

Reply:
Another option, to avoid adding another parameter to one of the main methods schema writers may call, to add another function named something like `readUnverifiedDataFromAsResult`. I also don't see a good way to do this without a more major refactor, so either way I think this is fine for version 1.